### PR TITLE
Extend precedence to lint for masking bits and shifts

### LINKED ISSUE #6632: Lint masking bits and shift
### What it does

Consider this code snippet:
```rust
pub const fn to_4_be_nibles(op: u16) -> [u8; 4] {
    [
        (op & 0xF000 >> 12) as u8,
        (op & 0x0F00 >> 8) as u8,
        (op & 0x00F0 >> 4) as u8,
        (op & 0x000F) as u8,
    ]
}
``` 

We'd realize that the code is wrong. `>>` has [stronger precedence](https://doc.rust-lang.org/nightly/reference/expressions.html#expression-precedence) than `&` operator.
So we has to use `()` to group `(op & 0F000)` and the like.
I'd like to have a lint about this.

### Categories (optional)

- Kind: clippy::correctness

### Drawbacks

None.

### PR DESCRIPTION
**Pull request by goth-turtle**
Currently this lints for any expression `_ & _ << _` or `_ & _ >> _`, and suggests adding parentheses, like so: `(_ & _) << _`.
Also, im not really sure about my wording with the documentation i added, so id appreciate some feedback on that. Thanks!

changelog: extended [`precedence`] to lint for bitmasking with shifting without parentheses

fixes #6632

### CODE REVIEW

**Comment by Manishearth** [comment]
r? @Jarcho 

(I'm travelling rn and might not be able to do full reviews, can rubber-stamp after you've had a look at it)

**Comment by Jarcho** [comment]
You'll need to switch the suggestion to `x & (y >> z)`? The precedence lint is meant as a note that the code as written may be easy to misread, not that the code is necessarily wrong.

I'll give a proper review tomorrow.

File: clippy_lints/src/precedence.rs [hunk 1]
```diff
@@ -56,86 +58,118 @@ impl EarlyLintPass for Precedence {
...
+                sugg,
+                appl,
+            );
+        };
+
+        if !is_bit_op(op) {
+            return;
+        }
+        let mut applicability = Applicability::MachineApplicable;
+        match (is_arith_expr(left), is_arith_expr(right)) {
```
Comment (Jarcho) [inline clippy_lints/src/precedence.rs:84]: This is the match expression that would need to change.

**Review by Jarcho — CHANGES_REQUESTED** [review]
Finished an initial read of the changes. The current implementation will lint twice on expression such as `a + b & x >> d`.

I'd also like to see this extended to handle any combination of `&`, `|`, and `^` with either shift operation. The approach would be to change the match expression to `match (op, get_bin_opt(left), get_bin_opt(right))` and then fill in arms like `(BitAnd | BitOr | BitXor, Some(Shl | Shr | Add | ...), Some(Shl | Shr | Add | ...))`. `get_bin_opt` will need to be written as well.

**Comment by giraffate** [comment]
@goth-turtle ping from triage. Can you have any question on this?

**Comment by blyxyas** [comment]
@goth-turtle closing this it has been inactive for 403 days. Closing this as (from your contribution history) it seems like you're no longer interested in OSS (nothing wrong about that :heart:).

If you return to the community, and are interested in continuing this PR, don't hesitate to reopen this PR (●ↀωↀ●)

**Pull request closed without merging**